{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cleaning the GWAS dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>...</th>\n",
       "      <th>CONTEXT</th>\n",
       "      <th>INTERGENIC</th>\n",
       "      <th>RISK ALLELE FREQUENCY</th>\n",
       "      <th>P-VALUE</th>\n",
       "      <th>PVALUE_MLOG</th>\n",
       "      <th>P-VALUE (TEXT)</th>\n",
       "      <th>OR or BETA</th>\n",
       "      <th>95% CI (TEXT)</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>CNV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14701</th>\n",
       "      <td>2018-04-26</td>\n",
       "      <td>25751624</td>\n",
       "      <td>Onengut-Gumuscu S</td>\n",
       "      <td>2015-04-01</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/25751624</td>\n",
       "      <td>Fine mapping of type 1 diabetes susceptibility...</td>\n",
       "      <td>Type 1 diabetes</td>\n",
       "      <td>6,683 European ancestry cases, 12,173 European...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.17</td>\n",
       "      <td>6.000000e-14</td>\n",
       "      <td>13.221849</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina [135870]</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14702</th>\n",
       "      <td>2018-04-26</td>\n",
       "      <td>25751624</td>\n",
       "      <td>Onengut-Gumuscu S</td>\n",
       "      <td>2015-04-01</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/25751624</td>\n",
       "      <td>Fine mapping of type 1 diabetes susceptibility...</td>\n",
       "      <td>Type 1 diabetes</td>\n",
       "      <td>6,683 European ancestry cases, 12,173 European...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.45</td>\n",
       "      <td>7.000000e-10</td>\n",
       "      <td>9.154902</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina [135870]</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19605</th>\n",
       "      <td>2019-08-07</td>\n",
       "      <td>31152121</td>\n",
       "      <td>Zhu M</td>\n",
       "      <td>2019-05-31</td>\n",
       "      <td>Diabetes Care</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31152121</td>\n",
       "      <td>Identification of Novel T1D Risk Loci and Thei...</td>\n",
       "      <td>Type 1 diabetes</td>\n",
       "      <td>1,005 Han Chinese ancestry cases, 1,257 Han Ch...</td>\n",
       "      <td>1,378 Han Chinese ancestry cases, 3,774 Han Ch...</td>\n",
       "      <td>...</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.28</td>\n",
       "      <td>3.000000e-08</td>\n",
       "      <td>7.522879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.26</td>\n",
       "      <td>[1.16-1.36]</td>\n",
       "      <td>Illumina [787224] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19607</th>\n",
       "      <td>2019-08-07</td>\n",
       "      <td>31152121</td>\n",
       "      <td>Zhu M</td>\n",
       "      <td>2019-05-31</td>\n",
       "      <td>Diabetes Care</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31152121</td>\n",
       "      <td>Identification of Novel T1D Risk Loci and Thei...</td>\n",
       "      <td>Type 1 diabetes</td>\n",
       "      <td>1,005 Han Chinese ancestry cases, 1,257 Han Ch...</td>\n",
       "      <td>1,378 Han Chinese ancestry cases, 3,774 Han Ch...</td>\n",
       "      <td>...</td>\n",
       "      <td>splice_region_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.46</td>\n",
       "      <td>2.000000e-232</td>\n",
       "      <td>231.698970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.28</td>\n",
       "      <td>[3.92-4.68]</td>\n",
       "      <td>Illumina [787224] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147068</th>\n",
       "      <td>2008-06-16</td>\n",
       "      <td>17632545</td>\n",
       "      <td>Hakonarson H</td>\n",
       "      <td>2007-07-15</td>\n",
       "      <td>Nature</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/17632545</td>\n",
       "      <td>A genome-wide association study identifies KIA...</td>\n",
       "      <td>Type 1 diabetes</td>\n",
       "      <td>467 European ancestry trios, 561 European ance...</td>\n",
       "      <td>390 European ancestry trios, 1,333 European an...</td>\n",
       "      <td>...</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.13</td>\n",
       "      <td>1.000000e-16</td>\n",
       "      <td>16.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.30</td>\n",
       "      <td>[6.97-9.89]</td>\n",
       "      <td>Illumina [543071]</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       DATE ADDED TO CATALOG  PUBMEDID       FIRST AUTHOR        DATE  \\\n",
       "14701             2018-04-26  25751624  Onengut-Gumuscu S  2015-04-01   \n",
       "14702             2018-04-26  25751624  Onengut-Gumuscu S  2015-04-01   \n",
       "19605             2019-08-07  31152121              Zhu M  2019-05-31   \n",
       "19607             2019-08-07  31152121              Zhu M  2019-05-31   \n",
       "147068            2008-06-16  17632545       Hakonarson H  2007-07-15   \n",
       "\n",
       "              JOURNAL                                  LINK  \\\n",
       "14701       Nat Genet  www.ncbi.nlm.nih.gov/pubmed/25751624   \n",
       "14702       Nat Genet  www.ncbi.nlm.nih.gov/pubmed/25751624   \n",
       "19605   Diabetes Care  www.ncbi.nlm.nih.gov/pubmed/31152121   \n",
       "19607   Diabetes Care  www.ncbi.nlm.nih.gov/pubmed/31152121   \n",
       "147068         Nature  www.ncbi.nlm.nih.gov/pubmed/17632545   \n",
       "\n",
       "                                                    STUDY    DISEASE/TRAIT  \\\n",
       "14701   Fine mapping of type 1 diabetes susceptibility...  Type 1 diabetes   \n",
       "14702   Fine mapping of type 1 diabetes susceptibility...  Type 1 diabetes   \n",
       "19605   Identification of Novel T1D Risk Loci and Thei...  Type 1 diabetes   \n",
       "19607   Identification of Novel T1D Risk Loci and Thei...  Type 1 diabetes   \n",
       "147068  A genome-wide association study identifies KIA...  Type 1 diabetes   \n",
       "\n",
       "                                      INITIAL SAMPLE SIZE  \\\n",
       "14701   6,683 European ancestry cases, 12,173 European...   \n",
       "14702   6,683 European ancestry cases, 12,173 European...   \n",
       "19605   1,005 Han Chinese ancestry cases, 1,257 Han Ch...   \n",
       "19607   1,005 Han Chinese ancestry cases, 1,257 Han Ch...   \n",
       "147068  467 European ancestry trios, 561 European ance...   \n",
       "\n",
       "                                  REPLICATION SAMPLE SIZE  ...  \\\n",
       "14701                                                 NaN  ...   \n",
       "14702                                                 NaN  ...   \n",
       "19605   1,378 Han Chinese ancestry cases, 3,774 Han Ch...  ...   \n",
       "19607   1,378 Han Chinese ancestry cases, 3,774 Han Ch...  ...   \n",
       "147068  390 European ancestry trios, 1,333 European an...  ...   \n",
       "\n",
       "                                   CONTEXT INTERGENIC RISK ALLELE FREQUENCY  \\\n",
       "14701                       intron_variant        0.0                  0.17   \n",
       "14702                       intron_variant        1.0                  0.45   \n",
       "19605   non_coding_transcript_exon_variant        0.0                  0.28   \n",
       "19607                splice_region_variant        0.0                  0.46   \n",
       "147068                  intergenic_variant        1.0                  0.13   \n",
       "\n",
       "              P-VALUE PVALUE_MLOG P-VALUE (TEXT) OR or BETA 95% CI (TEXT)  \\\n",
       "14701    6.000000e-14   13.221849            NaN       1.20           NaN   \n",
       "14702    7.000000e-10    9.154902            NaN       1.12           NaN   \n",
       "19605    3.000000e-08    7.522879            NaN       1.26   [1.16-1.36]   \n",
       "19607   2.000000e-232  231.698970            NaN       4.28   [3.92-4.68]   \n",
       "147068   1.000000e-16   16.000000            NaN       8.30   [6.97-9.89]   \n",
       "\n",
       "         PLATFORM [SNPS PASSING QC]  CNV  \n",
       "14701             Illumina [135870]    N  \n",
       "14702             Illumina [135870]    N  \n",
       "19605   Illumina [787224] (imputed)    N  \n",
       "19607   Illumina [787224] (imputed)    N  \n",
       "147068            Illumina [543071]    N  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "data = pd.read_csv('gwas_catalog_v1.0-associations_e108_r2023-01-14.tsv', sep='\\t', low_memory=False)\n",
    "#data =data.loc[data['DISEASE/TRAIT']=='Type 1 diabetes']   # select only rows regarding type 1 diabetes\n",
    "#data =data.loc[data['CHR_ID']=='6']                        # select only rows for chromosome 6\n",
    "data.to_csv('gwas_trimmed.tsv', sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8 (tags/v3.10.8:aaaf517, Oct 11 2022, 16:50:30) [MSC v.1933 64 bit (AMD64)]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "958ecc49f845cae67b922f9a0ace456f2aeb07b09c0ec55961dea76115aaae27"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
