% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%

%%% Setup %%%
\documentclass[12pt,a4paper]{article}
\title{Project Documentation}
\title{\huge \textbf{Project Documentation}}
\author{Team Duck Duck Go}
\date{} 					% no date if empty, current date if commented out

\widowpenalty=10000	% prevents breaking before last line of a paragraph
\clubpenalty=10000	% prevents breaking after first line of a paragraph
\raggedbottom			% allows bottom of page to be natural, not flush

\usepackage{placeins} % allows for \FloatBarrier
\usepackage{tcolorbox}
\newtcolorbox{code}{colback=gray!5!white,% boxrule=0mm,
     grow to left by=20mm,
     grow to right by=20mm,
     sharp corners}

\usepackage{graphicx}
\graphicspath{{./images}}	% path where images are found
\usepackage{minted}

\newcommand{\mintfile}[1]{
\begin{tcolorbox}[colback=gray!5!white,% boxrule=0mm,
	grow to left by=20mm,
    grow to right by=20mm,
    sharp corners]{{    \small \inputminted[breaklines]{python}{#1}		}}
\end{tcolorbox}}


\newcommand{\mintfileTitle}[2]{
\begin{tcolorbox}[colback=gray!5!white,% boxrule=0mm,
	grow to left by=20mm,
    grow to right by=20mm,
    sharp corners,
    title=#2    ]{{    \small \inputminted[breaklines]{python}{#1}		}}
\end{tcolorbox}}



\newcommand{\sect}[1]{
\clearpage
\hypertarget{#1}{
\section{#1}\label{#1}}
}

\newcommand{\subsect}[1]{
\FloatBarrier % Prevents figures from previous sections entering this section
\hypertarget{#1}{
\subsection{#1}\label{#1}}
}

\newcommand{\subsubsect}[1]{
\hypertarget{#1}{
\subsection{#1}\label{#1}}
}


\usepackage{caption}
\newcommand{\pic}[2]{
\begin{figure}[h]
    \centering
    \captionsetup{justification=centering}
    \includegraphics[width=\textwidth]{#1}
    \caption{#2}
    \label{#1}
\end{figure}
}

%%% End my setup %%%

%%%%% Auto generated by pandoc %%%%%

\usepackage{amsmath,amssymb}
\usepackage{lmodern}
\usepackage{iftex}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{textcomp} % provide euro and other symbols

% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}

\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same} % disable monospaced font for URLs
\hypersetup{
  hidelinks,
  pdfcreator={LaTeX via pandoc}}


%%%%% End pandoc stuff %%%%%

\begin{document}

\maketitle
\thispagestyle{empty}
\clearpage
\tableofcontents
\clearpage

\sect{Introduction}


This web application prototype is designed to retrieve information on
Single Nucleotide Polymorphisms (SNPs) seen in Type 1 Diabetes patients
identified by Genome wide association studies (GWAS). The database will
use information from the GWAS catalogue, along with population data from
Ensembl and the 1000 Genomes Project and functional information and Gene
Ontology information obtained through Ensembl's VEP tool which is all is
retrievable through a user friendly interface through the input of an
rsID, Chromosome position or a Gene name. The site also allows the user
to calculate Linkage Disequilibrium (LD) of SNPs selected for each
population producing a text file containing the LD values and plot these
values as a LD heatmap. The user is also able to enter multiple SNPs and
return a Manhattan plot of the p-values.

\subsect{Prerequisites:}


\subsubsect{Functions used throughout:}

A number of functions were used throughout the code such as\ldots{}

\mintfile{code_snippets/placeholder.py}

\sect{Structure}
\pic{structure}{Structure of the data flow}

\subsect{GWAS}

This information was downloaded from the GWAS catalogue where a TSV file was downloaded and then trimmed:

\mintfile{code_snippets/placeholder.py}
This code uses pandas to open the TSV file, creates a dataframe called data, and removes any special characters from the column names, as SQL does not interact with special characters very well.

\mintfile{code_snippets/placeholder.py}
The data frame is then filtered further so that it only has data that references T1D in the “disease\_trait” column so that only T1D data remains in the dataframe. Next all SNPs that don't have rsIDs are removed, as some cells had incompatible data in this column.

\mintfile{code_snippets/placeholder.py}
The dataframe is then trimmed again so that it contains only the columns of interest   "snps", "region", "chr\_pos", "chr\_id", "p\_value", "mapped\_gene".

\mintfile{code_snippets/placeholder.py}
Duplicate mapped\_gene information is then removed.

\mintfile{code_snippets/placeholder.py}
Next the column 'snps' was renamed to 'rsid' and made the change directly to the dataframe by setting inplace=True.


\subsect{Functional information and Gene Ontology}

We have used Ensembl's Variant Effect Predictor web tool to gather the Functional and Ontology data by
submitting a job with the rsIDs as input from the above GWAS TSV file.
After running the job, we get an output of text file with columns like:

\begin{itemize}

\item rsid: the reference SNP identifier for the variant.
\item Allele: the alternative allele observed at the variant site.
\item impact: the impact of the variant on the affected gene
\item consequence: the consequence of the variant on the affected gene
\item location: the location of the variant within the affected gene
\item Gene: the Ensembl gene ID of the affected gene.
\item Symbol: the gene symbol or name.
\item Feature\_Type: the type of genomic feature the variant is located in
\item Feature: the Ensembl ID of the specific feature the variant is located in
\item Exon: the exon number(s) affected by the variant.
\item Intron: the intron number(s) affected by the variant.
\item HGVSc: the HGVS nomenclature for the variant at the cDNA .
\item HGVSp: the HGVS nomenclature for the variant at the protein level.
\item cDNA\_position: the position of the variant within the cDNA sequence of the affected gene.
\item CDS\_position: the position of the variant within the coding sequence of the affected gene.
\item Protein\_position: the position of the variant within the protein sequence of the affected gene.
\item Amino\_acids: the amino acid change resulting from the variant.
\item Codons: the DNA codon change resulting from the variant.
\item Existing\_variation: additional identifiers for the variant in other databases.
\item Distance: the distance to the nearest feature in the same or opposite strand.
\item Strand: the genomic strand the variant is located on.
\item FLAGS: additional information about the variant .
\item SYMBOL\_SOURCE: the source of the gene symbol or name.
\item HGNC\_ID: the HGNC gene ID of the affected gene.
\item MANE\_SELECT: indication of whether this transcript is the MANE (Matched Annotation from NCBI and EMBL-EBI) Select transcript.
\item MANE\_PLUS\_CLINICAL: indication of whether this transcript is the MANE Select Plus Clinical (MPC) transcript.
\item TSL: transcript support level (a measure of transcript annotation confidence).
\item APPRIS: annotation of principal isoforms for each gene.
\item ENSP: the Ensembl protein ID of the affected protein.
\item SIFT: prediction of the effect of the variant on protein function.
\item PolyPhen: prediction of the effect of the variant on protein function.
\item CLIN\_SIG: clinical significance of the variant.
\item SOMATIC: indication of whether the variant is somatic or germline.
\item PHENO: phenotype association of the variant.
\item PUBMED: PubMed ID of publications reporting functional evidence of the variant.
\item MOTIF\_NAME: name of the DNA motif affected by the variant.
\item MOTIF\_POS: position of the variant within the affected DNA motif.
\item HIGH\_INF\_POS: indication of whether the variant falls in a highly conserved position within the DNA motif.
\item MOTIF\_SCORE\_CHANGE: the effect of the variant on the score of the affected DNA motif.
\item TRANSCRIPTION\_FACTORS: transcription factors that bind the affected DNA motif.
\item CADD\_PHRED: Phred-scaled CADD score (Combined Annotation-Dependent Depletion), which predicts the deleteriousness of variants.
\item CADD\_RAW: the raw CADD score, which is a measure of the deleteriousness of variants.
\item GO Terms: Gene Ontology (GO) terms associated with the affected gene.

\end{itemize}

The VEP file provides detailed information about the functional and ontological consequences of genetic variants, including their impact on genes, proteins, and pathways.

We have converted the text file to a tsv file and trimmed down the file to include only the rsID, Alleles, CADD\_PHRED and CADD\_RAW scores columns for the functional data and the rsID, location, gene, symbol, GO terms columns for the Ontology data. We have further used these TSV files in our database.

CADD (Combined Annotation Dependent Depletion) is a tool used for predicting the potential harm caused by genetic variants. The tool generates a score that indicates the likelihood of a variant being deleterious.

One advantage of CADD over SIFT and PolyPhen is that CADD integrates a larger and more diverse set of functional annotations. It also considers the effects of variants on non-coding regions of the genome, which can be important for understanding the functional consequences of variants that are not in protein-coding regions.

\sect {Linkage Disequilibrium}

Linkage disequilibrium (LD) is the degree of non-random association of the allele of one SNP with the allele of another SNP within a population. LD is typically measured by two metrics: $D’$ and $r^2$.

$D’$ is the normalised values of $D$, the coefficient of linkage disequilibrium, where A and B are alleles of two SNPs in different loci:

\[D=P_A -  P_A P_B\]

$r^2$  is the correlation coefficient between two loci:

\[r^2 = \frac{D^2}{p_A(1-p_A)p_B(1-p_B)} \]


\subsect{Collecting data}

Linkage disequilibrium data was obtained from LDlink using the LDmatrix tool. $D’$ and $r^2$ values for SNPs were calculated using 1000 Genomes Project data for all three populations. LD data was obtained by inputting a list of SNPs from the same chromosome and selecting the population which would be used for allele frequency data for LD calculations. LDmatrix would produce two text files containing a matrix of results for $D’$ and $r^2$ values calculated between all SNPs pair combinations in the input list. This was performed separately for each population. Some SNPs did not have any LD data due to a lack of allele frequency data for those SNPs in the 1000 Genomes Project.

LD datasets containing $D’$ and $r^2$  values for Finnish, Toscani and British populations are loaded in with pandas as separate dataframes. Each dataframe has their index set to the first column which contains SNP rsIDs.
\mintfile{code_snippets/placeholder.py}

This function uses the itertools combination function to create a list of tuples containing all unique pairs of SNPs possible from a list of SNPs. The list is then separated into two lists containing the first and second element of each tuple.
\mintfile{code_snippets/placeholder.py}

An empty dataframe is created to be filled with rows containing data from all six dataframes. This loop uses the two lists created from the SNP list to index each dataframe and extract the respective LD value. These are used to create a list which is converted into a single row pandas dataframe which is added to the empty dataframe using pandas concat until data for all relevant pairwise LD calculations have been added. The completed dataframe is then outputted as a TSV file.
\mintfile{code_snippets/placeholder.py}


\subsect{Outputting LD results}
When a user searches by gene name or chromosomal coordinates, if multiple SNPs are returned, a list of SNPs is used to filter the LD dataset for all rows with entries for all pairwise LD calculations of SNPs in the list and output a results dataframe.


Before filtering, the list is checked for any SNPs which are not in the LD dataset due to lack of LD data and any offending SNPs are removed from the list.
\mintfile{code_snippets/placeholder.py}

The SNP list is then used to create two lists containing the first and second element of each tuple using the SNP\_pair\_lists() function defined earlier.
\mintfile{code_snippets/placeholder.py}


The LD dataset containing all available data for pairwise LD calculations is loaded in with pandas and an empty dataframe is created for the filtered data. The pair of SNP lists are then used to index the LD dataset dataframe for all rows with pairs of SNPs relevant to the user’s search query which are added to the LD results dataframe using pandas concat.
\mintfile{code_snippets/placeholder.py}

\subsect{LD heatmap plots}

When a user searches by gene name or chromosomal coordinates, if multiple SNPs are returned, a list of SNPs is also used to extract LD values for all relevant pairwise SNP calculations to create a dataframe used to create a heatmap plot of LD values.

The LD dataset is loaded in with pandas and SNPs not present in the dataset are removed from the list of SNPs passed from the user query. The SNP list is then used to create two lists containing the first and second element of each tuple using the SNP\_pair\_lists() function defined earlier.
\mintfile{code_snippets/placeholder.py}

An empty dataframe is created to be filled with LD values used to create the LD plot. The pair of SNP lists are then used to index the LD dataset dataframe and extract the LD value for all possible pairwise LD calculations from the SNP list. A row of LD values is created for each SNP where each column corresponds with the pairwise LD calculation with one of the SNPs from the list. Each row is added to the empty dataframe using pandas concat.
\mintfile{code_snippets/placeholder.py}

The LD matrix dataframe is passed to the ld\_plot function. The number of rows (n) is used to create a mask which will hide half of the heatmap to create a triangular plot. A coordinate matrix is also created to rotate the heatmap plot. The SNP list is used to create the axis labels located at the bottom of the plot. The function’s title parameter passes a string which is used to determine the plot title.
\mintfile{code_snippets/placeholder.py}

\sect{Manhattan Plot}

A Manhattan plot is a type of scatter graph which displays P values of the entire Genome-wide association study (GWAS) on genomic scale. The P-values are shown in genomic order by chromosomal position on the $x$-axis. The $y$-axis shows the $-log10$ value of the P-value for each polymorphism.

A TSV file containing all SNPs in type 1 diabetes was produced. The Manhattan plot needs the P-value of a SNP and the chromosomal position. Using pandas, the $-logp$ and cumulative positions were generated.

\mintfile{code_snippets/manPlot/createTSV.py}
The original file `GWAS\_T1D.tsv’ is used to create an additional column for -logp values and creating a column for cumulative positions by using the chromosome and chromosomal position values.
The cumulative position is needed to show the position of the SNP in the entire genome, rather than its position in the chromosome.

\mintfile{code_snippets/manPlot/createManPlot.py}
Bokeh tools was used to make the graph user friendly. Circle plots were chosen for easy visibility and selecting of each SNP position. Hover feature was added to allow the user to hover mouse over each plot, which turns it green, and shows the rs ID value next to the chromosomal position. The select feature allows users to select 1 or more plots by either clicking one plot, or holding shift and clicking plots for multiple plots, or using the box select tool to select plots in an entire section, and turns the plots purple. The zoom features allow the user to zoom into the graph and view plots that are close to each other or seem to be overlapping for more visual clarity. The reset and undo buttons allows the user to either go back to the original plot, or undo the previous action they had committed. The save tool will save the graph as a png file with the title included. The plots were made slightly transparent to make overlapping SNPs more visible.

\pic{manplot1}{All GWAS SNPs found in Type 1 Diabetes shown in a Manhattan Plot.
Users can view all SNPs for T1D, and plot the SNPs that they have searched for.
This feature is available when multiple SNPs are retrieved for the region searched.}

\pic{manplot2}{All GWAS SNPs for Type 1 Diabetes, Chromosome 6 only}

\sect{Website}

\subsect{Flask}

The main page was created using the following function. By adding the \verb|@app.route()| decorator,
the flask module can run the function whenever the root page is accessed.
The function creates an instance of class \texttt{QueryForm}, which it uses to receive user input.
The input is then formatted to better be understood by future functions,
before the parsed request and the request type are passed to the next page as the subpage and URL parameters, respectively.

\mintfile{code_snippets/flask/root.py}

The following function first receives the \texttt{SNP\_req} by reading the subpage name,
and the \texttt{req\_type} by making a \texttt{HTTP GET} request to retrieve the relevant argument.
If the request type is set to automatically detect, the request is tested against a series of
regex searches to determine the intended request type.

The \texttt{DBreq()} function is then called to retrieve the required information via SQL commands. This function will be elaborated upon in the following section.

\mintfile{code_snippets/flask/snp.py}

\subsect{SQLite}

To create the database, the \texttt{.to\_SQL()} method from Pandas was used, which receives a dataframe of a TSV file,
as well as an SQLite3 connection, and produces an SQLite compatible \texttt{.db} file

\mintfile{code_snippets/SQL/create_db.py}
The above code makes use of various functions from the \texttt{db\_scripts.py} custom library:

\mintfile{code_snippets/SQL/pdDB.py}

After connecting to the database file, rsIDs are extracted from the user's request,
if necessary via cross-referencing with with GWAS dataset.

\mintfile{code_snippets/SQL/DBreq1.py}

The resultant list of SNPs is then iterated over to fetch each corresponding entry from the
GWAS, Population, Functional, and Ontology tables.
Each entry is then stored in a dictionary to be passed to the main flask code:

\FloatBarrier

\mintfile{code_snippets/SQL/DBreq2.py}
%\thispagestyle{empty} % remove page number
\FloatBarrier

\sect{Navigation}
Screenshot how site works how to navigate\ldots








\sect{User Feedback}
10 biologists, ages 25-60, were given the opportunity to use the software to search for either a SNP, region or gene. They were then asked to either retrieve GO data, Linkage disequilibrium plot, or the Manhattan plot.

Items that were improved after user feedback:

\begin{itemize}
\item Colours/Themes: users found it easier to read with a darker theme; background was made darker, themes added.
\item Font size/style: users found the font easy to read and see, including users with visual impairments.
\item Quick links: users did not like scrolling through to find the data they were looking for; adding the quick links makes it quicker to get to the terms.
\item Return home button: this feature was added to all pages to allow users to return to the main page.
\item Overall users found the genome browser very user friendly and easy to use and enjoyed all the interactive features. They said they liked how easy it is to search and retrieve information, and enlightened them on the vast amount of genome data we now have access to through GWAS.
\end{itemize}

Overall users found the genome browser very user friendly and easy to use and enjoyed all the interactive features. They said they liked how easy it is to search and retrieve information, and enlightened them on the vast amount of genome data we now have access to through GWAS.

\sect{References}




\end{document}





















